File: CMPS 101 - Programming Assignment 2/Lex.c
// Lex.c
// Andrea Vitek
// 1559060
// pa 2
// This program reads strings from a text file, sorts them using a separate List ADT
// and exports the sorted strings to a text file

#include<stdio.h>
#include<stdlib.h>
#include <string.h>
#include "List.h"

#define MAX_LEN 160

// Forward declarations
void insertionSortForList(char** inputListStrings, int inputListStringsSize, List inputListIndexes);
char** getInputStrings(char filename[], int* arraySize);
void printListInOrder(char** inputListStrings, List inputListIndexes);
void writeOutputFile(char filename[], char** inputListStrings, List inputListIndexes);

// Main function
int main(int argc, char * argv[]) {
    // Check command lines for correct number of arguments
    if (argc != 3) {
        printf("Usage: %s <input file> <output file>\n", argv[0]);
        exit(1);
    }

    int* arraySize = malloc(sizeof(int));
    char** stringsArray = getInputStrings(argv[1], arraySize);
    List inputListIndexes = newList();

    insertionSortForList(stringsArray, *arraySize, inputListIndexes);

    writeOutputFile(argv[2], stringsArray, inputListIndexes);

    for (int i = 0; i < *arraySize; ++i) {
        free(stringsArray[i]);
    }

    free(arraySize);
    free(stringsArray);
    freeList(&inputListIndexes);

    return(0);
}

// insertionSortForList()
// Performs insertionSort to sort the indexes of a String array in alphabetical order by storing
// the array's sorted indexed in a List ADT
void insertionSortForList(char** inputListStrings, int inputListStringsSize, List inputListIndexes) {
    append(inputListIndexes, 0);

    for (int j = 1; j < inputListStringsSize; j++) {
        moveFront(inputListIndexes);

        while (index(inputListIndexes) >= 0) {
            int cursor = get(inputListIndexes);

            if (strcmp(inputListStrings[j], inputListStrings[cursor]) < 0) {
                insertBefore(inputListIndexes, j);
                break;
            }

            moveNext(inputListIndexes);

            if (index(inputListIndexes) == -1) {
                append(inputListIndexes, j);
            }
        }
    }
}

// getInputStrings()
// Returns an array of strings from an input file
char** getInputStrings(char filename[], int* arraySize) {
    int count = 0;
    FILE *in;
    char line[MAX_LEN];
    char tokenlist[MAX_LEN];
    char* token;

    in = fopen(filename, "r");

    if (in == NULL) {
        //printf("Unable to open file %s for reading\n", argv[1]);
        printf("Unable to open file for reading\n");
        exit(1);
    }

    // Count the number of strings in the input file
    while (fgets(line, MAX_LEN, in) != NULL) {
        token = strtok(line, " \n");
        tokenlist[0] = '\0';

        while (token != NULL) {
            strcat(tokenlist, token);
            token = strtok(NULL, "\0");

            // Only count if the line is not empty
            if (strcmp(tokenlist, "") != 0) {
                count++;
            }
        }
    }

    // Reset the input file reading cursor
    rewind(in);

    char** stringsArray = malloc(count * sizeof(char*));

    for (int i = 0 ; i < count; ++i) {
        stringsArray[i] = malloc(count * sizeof(char));
    }

    int i = 0;

    while (fgets(line, MAX_LEN, in) != NULL) {
        token = strtok(line, " \n");
        tokenlist[0] = '\0';

        while (token != NULL) {
            strcat(tokenlist, token);
            token = strtok(NULL, "\0");

            if (strcmp(tokenlist, "") != 0) {
                strcpy(stringsArray[i], tokenlist);
            }

            i++;
        }
    }

    fclose(in);

    *arraySize = count;

    return stringsArray;
}

// printListInOrder()
// Prints the Strings of an array in the correct alphabetical order to the console
// following their sorted indexes in a List ADT
void printListInOrder(char** inputListStrings, List inputListIndexes) {
    moveFront(inputListIndexes);

    while (index(inputListIndexes) >= 0) {
        printf("%s\n", inputListStrings[get(inputListIndexes)]);
        moveNext(inputListIndexes);
    }
}

// writeOutputFile()
// Prints the Strings of an array in the correct alphabetical order to an output file
// following their sorted indexes in a List ADT
void writeOutputFile(char filename[], char** inputListStrings, List inputListIndexes) {
    FILE * out = fopen(filename, "w");

    if (out == NULL) {
        printf("Unable to open file \"%s\" for writing\n", filename);
        exit(1);
    }

    moveFront(inputListIndexes);

    while (index(inputListIndexes) >= 0) {
        fprintf(out, "%s\n", inputListStrings[get(inputListIndexes)]);
        moveNext(inputListIndexes);
    }

    fclose(out);
}

File: CMPS 101 - Programming Assignment 1/Lex.java
// Lex.java
// Andrea Vitek
// 1559060
// pa 1
// This program reads strings from a text file, sorts them using a separate List ADT
// and exports the sorted strings to a text file

import java.io.*;
import java.util.Scanner;

public class Lex {	
	public static void main(String[] args) {
		//String[] files = Lex.getIOFiles();
		String[] files = args;
		
		if (files.length != 2) {
			System.out.println("You did not enter two files names separated by a space.");
			System.exit(0); // Exit as requested
		}

		String[] inputListStrings;
		List inputListIndexes = new List();

		try {
			inputListStrings = Lex.loadFromTextFile(files[0]);
			Lex.insertionSortForList(inputListStrings, inputListIndexes);
			
			Lex.exportToTextFile(inputListStrings, inputListIndexes, files[1]);
			
			System.out.println("Completed");

		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	// Performs insertionSort to sort the indexes of a String array in alphabetical order by storing
	// the array's sorted indexed in a List ADT
	public static void insertionSortForList(String[] inputListStrings, List inputListIndexes) {
		inputListIndexes.append(0);
		
		for (int j = 1; j < inputListStrings.length; j++) {
			inputListIndexes.moveFront();
			
			while (inputListIndexes.index() >= 0) {
				int cursor = inputListIndexes.get();
				
				if (inputListStrings[j].compareTo(inputListStrings[cursor]) < 0) {
					inputListIndexes.insertBefore(j);
					break;
				}
				
				inputListIndexes.moveNext();
				
				if (inputListIndexes.index() == -1) {
					inputListIndexes.append(j);
				}
			}
		}
	}

	// Asks for the input and output file names to the user
	public static String[] getIOFiles() {
		Scanner scanner = new Scanner(System.in);
		String[] IOFiles;

		// Get the user input
		System.out.println("Enter the name of the input and output files, in that order and including the file extension, separated by a space.");

		String line = scanner.nextLine();

		IOFiles = line.split(" ");

		if (IOFiles.length != 2) {
			System.out.println("You did not enter two files names separated by a space.");
			System.exit(0); // Exit as requested
		}

		scanner.close();

		return IOFiles;
	}

	// Loads the lines of an input file into a String array.
	public static String[] loadFromTextFile(String inputFileURL) throws IOException {
		File f = new File(inputFileURL);
		String[] inputListStrings = null;

		if (!f.exists()) {
			throw new RuntimeException("File does not exist");
		}

		try {
			// Count number of lines in input file
			LineNumberReader  lnr = new LineNumberReader(new FileReader(new File(inputFileURL)));
			lnr.skip(Long.MAX_VALUE);
			int totalLines = lnr.getLineNumber() + 1;
			lnr.close();

			inputListStrings = new String[totalLines];
			int currentIndex = 0;

			try (BufferedReader br = new BufferedReader(new FileReader(inputFileURL))) {
				String currentLine;

				while ((currentLine = br.readLine()) != null) {
					inputListStrings[currentIndex] = currentLine;
					currentIndex++;
				}
			}

		} catch (IOException e) {
			System.out.println("Error! " + e.toString());
		}

		return inputListStrings;
	}
	
	// Exports the Strings of an array in the correct alphabetical order to a text file
	// following their sorted indexes in a List ADT
	public static void exportToTextFile(String[] inputListStrings, List inputListIndexes, String outputFileName) {
		try (PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(outputFileName, true)))) {
			inputListIndexes.moveFront();

			while (inputListIndexes.index() >= 0) {
				out.print(inputListStrings[inputListIndexes.get()] + "\n");
				inputListIndexes.moveNext();
			}

		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	// Prints the Strings of an array in the correct alphabetical order to the console
	// following their sorted indexes in a List ADT
	public static void printStringsInOrder(String[] inputListStrings, List inputListIndexes) {
		inputListIndexes.moveFront();

		System.out.print("[ ");

		while (inputListIndexes.index() >= 0) {
			System.out.print(inputListStrings[inputListIndexes.get()] + " ");
			inputListIndexes.moveNext();
		}

		System.out.print("]");
	}
}